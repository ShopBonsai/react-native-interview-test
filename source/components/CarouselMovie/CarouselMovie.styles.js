import styled from 'styled-components'
import { View, FlatList, Text } from 'react-native'

export const CarouselMovieContainer = styled(View)`
  flex-direction: column;
  padding: 10px;
`
export const CarouselMovieContent = styled(FlatList)``

export const CarouselMovieTitle = styled(Text)`
  color: white;
  font-size: 20px;
  padding-bottom: 15px;
`
