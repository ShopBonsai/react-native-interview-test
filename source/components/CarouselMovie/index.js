import React from 'react'
import PropTypes from 'prop-types'
import {
  CarouselMovieContainer,
  CarouselMovieContent,
  CarouselMovieTitle
} from './CarouselMovie.styles'
import CardMovie from '../CardMovie'


/**
 * The CarouselMovie it's responsible by
 * to render a list of  item on vertical position
 *
 * @export {function}
 * @returns {JSX.Element}
 */
export default function CarouselMovie(props) {
  const { items, title } = props

  // Main return
  return (
    <CarouselMovieContainer>
      <CarouselMovieTitle>{title}</CarouselMovieTitle>
      <CarouselMovieContent
        horizontal
        showsHorizontalScrollIndicator={false}
        data={items}
        renderItem={({ item }) => item }
      />
    </CarouselMovieContainer>
  )
}

// default values
CarouselMovie.defaultProps = {
  items: [],
  title: 'Title carousel'
}

// prop validation
CarouselMovie.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({ type: CardMovie })),
  title: PropTypes.string
}
