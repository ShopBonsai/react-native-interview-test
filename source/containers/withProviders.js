import { compose } from "recompose"

import reduxStore from "../store"

import withRedux from "./withRedux"
import withSafeArea from "./withSafeArea"

export default compose(withRedux(reduxStore), withSafeArea)
