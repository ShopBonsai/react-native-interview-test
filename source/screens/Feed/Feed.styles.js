import styled from 'styled-components'
import { View, FlatList, Text } from 'react-native'


export const FeedContainer = styled(View)`
  flex: 1;
  background-color: #30312c;
`
export const FeedContent = styled(FlatList)`
  flex: 1;
  height: 100%;
`

export const FeedEmptyList = styled(Text)`
  color: #ffffff;
  font-size: 30px;
  align-self: center;
  margin: 85% auto 0 auto;
`
