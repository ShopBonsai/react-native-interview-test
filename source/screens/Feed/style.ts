import styled from 'styled-components/native';

export const Container = styled.View`
  flex: 1;
  background-color: #2d3144;
  padding: 0 10px;
`;
