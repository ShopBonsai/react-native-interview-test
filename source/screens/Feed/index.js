// libraries
import React, { useEffect, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'

// components
import CardMovie from '../../components/CardMovie'
import CarouselMovie from '../../components/CarouselMovie'

// actions
import { fetchMovies } from '../../redux/actions/movies'

// styles
import { FeedContainer, FeedContent, FeedEmptyList } from './Feed.styles'
import LoadingSpinner from '../../components/LoadingSpinner'


/**
 * The renderCarouselItems it's
 * responsible by to render a array of cardmovie
 *
 * @param {Array} item
 * @returns {Array(JSX.Element)}
 */
export function renderCarouselItems(item = null) {
  if (!item || item.length === 0) return null
  return item.map(({ image, title, id }) => (
    <CardMovie
      key={id}
      title={title}
      thumbnail={image}
    />
  ))
}
/**
   * The renderCarousel it's responsible by to render
   * a carousel by genre
   *
   * @param {object} { item }
   * @returns {JSX.Element}
   */
export function renderCarousel({ item } = {}) {
  if (!item) return null
  return item && (
    <CarouselMovie
      key={item.key.toString()}
      title={item.genre}
      items={renderCarouselItems(item.items)}
    />
  )
}

/**
 * The Feed it's function component responsible by
 * to render the screen feed
 *
 * @export {Fucntion}
 * @returns {JSX.Element}
 */
export default function Feed() {
  // redux by hooks
  const { data, isLoading, noMoreItems } = useSelector(({ movieReducer }) => movieReducer)
  const dispatch = useDispatch()
  const [skip, setSkip] = useState(0)
  /** **
   *  useEffect as "did mount"
   */
  useEffect(() => {
    dispatch(fetchMovies(skip, 15))
  }, [skip])

  // Load more items
  const loadMore = () => (!noMoreItems && !isLoading ? setSkip(skip + 1) : null)

  // Main return
  return (
    <FeedContainer>
      <FeedContent
        keyExtractor={({ key }) => key.toString()}
        data={data}
        renderItem={item => renderCarousel(item)}
        onEndReached={() => loadMore() }
        onEndReachedThreshold={10}
        onRefresh={() => setSkip(0)}
        initialNumToRender={15}
        refreshing={isLoading && data.length > 0}
        ListEmptyComponent={ () => (
          !isLoading && data.length === 0 ? <FeedEmptyList>Oops! No data :/ </FeedEmptyList> : null
        )}
      />
      { isLoading && data.length === 0 && <LoadingSpinner />}
    </FeedContainer>
  )
}
