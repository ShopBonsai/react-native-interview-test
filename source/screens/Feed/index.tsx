import React from 'react';
import { NavigationScreenComponent } from 'react-navigation';

import FeedMovieList from '../../containers/FeedMovieList';

import { Container } from './style';

const Feed: NavigationScreenComponent = () => {
  return (
    <Container testID="feed-screen">
      <FeedMovieList />
    </Container>
  );
};

export default Feed;
