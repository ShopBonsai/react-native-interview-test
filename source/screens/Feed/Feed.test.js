// libraries
import React from 'react'
import renderer from 'react-test-renderer'

// components
import Feed, { renderCarousel, renderCarouselItems } from './index'
import withProviders from '../../containers/withProviders'
import CarouselMovie from '../../components/CarouselMovie'

describe('Feed Screen', () => {
  it('Renders Feed correctly', () => {
    const Component = withProviders(Feed)
    const tree = renderer
      .create(<Component />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Test the renderCarousel of feed screen when the params are empty', async () => {
    expect(renderCarousel()).toBe(null)
  })
  it('Test the renderCarouselItems of feed screen when the params are empty', async () => {
    expect(renderCarouselItems()).toBe(null)
  })

  it('Test the renderCarousel of feed screen with the correct parms', async () => {
    const item = {
      genre: 'genre',
      key: 'key',
      items: []
    }
    const Carousel = renderCarousel({ item })
    expect(Carousel).toMatchSnapshot()
    expect(Carousel.type(item)).toBeInstanceOf(Object)
    expect(Carousel.type).toEqual(CarouselMovie)
  })
})
