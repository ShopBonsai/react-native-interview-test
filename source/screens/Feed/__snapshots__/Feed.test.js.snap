// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed Screen Renders Feed correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "black",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    className="sc-bZQynM wqEUz"
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      className="sc-gzVnrw idEVjV"
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={15}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={10}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
    <View
      className="sc-dnqmqq fAyWKE"
    >
      <ActivityIndicator
        animating={true}
        className="sc-iwsKbI dNpkGG"
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />
      <Text
        className="sc-gZMcBi gfnEmy"
      >
        Loading...
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`Feed Screen Test the renderCarousel of feed screen with the correct parms 1`] = `
<CarouselMovie
  items={null}
  title="genre"
/>
`;
